### PR TITLE
Expose submodules in __init__.py for type checking

### LINKED ISSUE #47027: Fix the way imports are done to be more correct for static type checkers
## 🐛 Bug

There are a number of imports in torch of submodules that are implicitly reexported. E.g. you can do (and it is commonly done in examples):

```
import torch

torch.nn.<something>
```

In `__init__.py`, the submodules are imported as (using the `nn` example):

```
import torch.nn
```

But to be properly re-exported for static type checkers, this is not correct. Instead, something like:

```
import torch.nn as nn
```

is needed.

### PR DESCRIPTION
**Pull request by MHDante**
Draft submission, addresses #47027

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -567,33 +567,35 @@ def _assert(condition, message):
...
-import torch.distributions
-import torch.testing
+from torch import onnx as onnx
+from torch import jit as jit
+from torch import linalg as linalg
+from torch import hub as hub
+from torch import random as random
+from torch import distributions as distributions
+from torch import testing as testing
 import torch.backends.cuda
```
Comment (nurpax) [inline torch/__init__.py:593]: Backends are also accessed via `import torch` and used like `torch.backends.cuda.is_built()` (https://pytorch.org/docs/stable/backends.html).  I guess this change as it is now won't fix those imports for pyright?

I care _more_ about the top-level imports but it'd be bummer if pyright would work only for a subset of imports.

File: torch/__init__.py [hunk 2]
```diff
@@ -567,37 +567,45 @@ def _assert(condition, message):
...
-import torch.nn.intrinsic
-import torch.nn.quantizable
-import torch.nn.quantized
-import torch.optim
+from torch import fft as fft
+from torch import futures as futures
+from torch import nn as nn
+# import torch.nn.intrinsic # moved to torch.nn
+# import torch.nn.quantizable # moved to torch.nn
+# import torch.nn.quantized # moved to torch.nn
```
Comment (ezyang) [inline torch/__init__.py:580]: This movement is fine, you can remove this comments :)

File: torch/backends/__init__.py [hunk 3]
```diff
@@ -45,3 +45,12 @@ def __init__(self, m, name):
 
     def __getattr__(self, attr):
         return self.m.__getattribute__(attr)
+
+# Re-export submodules
+
+from torch.backends import cuda as cuda
+# from torch.backends import cudnn as cudnn # was missing from torch/__init__.py
```
Comment (ezyang) [inline torch/backends/__init__.py:52]: Based on discussion in the issue, OK to add this

File: torch/utils/__init__.py [hunk 4]
```diff
@@ -11,3 +11,10 @@ def set_module(obj, mod):
 
 #: Path to folder containing CMake definitions for Torch package
 cmake_prefix_path = _osp.join(_osp.dirname(_osp.dirname(__file__)), 'share', 'cmake')
+
+
+# Re-export submodules
+
+from torch.utils import backcompat as backcompat
+# cannot export data , as this would cause recursive import
+# from torch.utils import data as data
```
Comment (ezyang) [inline torch/utils/__init__.py:20]: But you still deleted it from the source module? Might be better to handle the recursive import, but that can be fixed later.

**Comment by ezyang** [comment]
Test failures are real. Unfortunately you're going to have to figure it out:

```

Jan 18 05:16:13 + python -c 'import torch; print(torch.__config__.show())'
Jan 18 05:16:13 Traceback (most recent call last):
Jan 18 05:16:13   File "<string>", line 1, in <module>
Jan 18 05:16:13   File "/opt/conda/lib/python3.6/site-packages/torch/__init__.py", line 531, in <module>
Jan 18 05:16:13     from .functional import *
Jan 18 05:16:13   File "/opt/conda/lib/python3.6/site-packages/torch/functional.py", line 6, in <module>
Jan 18 05:16:13     import torch.nn.functional as F
Jan 18 05:16:13   File "/opt/conda/lib/python3.6/site-packages/torch/nn/__init__.py", line 9, in <module>
Jan 18 05:16:13     from torch.nn import intrinsic as intrinsic
Jan 18 05:16:13   File "/opt/conda/lib/python3.6/site-packages/torch/nn/intrinsic/__init__.py", line 1, in <module>
Jan 18 05:16:13     from .modules import *
Jan 18 05:16:13   File "/opt/conda/lib/python3.6/site-packages/torch/nn/intrinsic/modules/__init__.py", line 1, in <module>
Jan 18 05:16:13     from .fused import _FusedModule
Jan 18 05:16:13   File "/opt/conda/lib/python3.6/site-packages/torch/nn/intrinsic/modules/fused.py", line 5, in <module>
Jan 18 05:16:13     class _FusedModule(torch.nn.Sequential):
Jan 18 05:16:13 AttributeError: module 'torch' has no attribute 'nn'
Jan 18 05:16:13 + cleanup
```

File: torch/__init__.py [hunk 5]
```diff
@@ -567,37 +567,45 @@ def _assert(condition, message):
...
+# import torch.backends.cuda # moved to torch.backends
+# import torch.backends.mkl # moved to torch.backends
+# import torch.backends.mkldnn # moved to torch.backends
+# import torch.backends.openmp # moved to torch.backends
+# import torch.backends.quantized # moved to torch.backends
+from torch import quantization as quantization
+import torch.utils.data # kept here to avoid recursive import of Tensor 
+
+# Kept private due to prefix.
 import torch.optim._multi_tensor
```
Comment (jakebailey) [inline torch/__init__.py:604]: This appears to be done above, if you need to remove that too.

File: torch/__init__.py [hunk 6]
```diff
@@ -567,37 +567,45 @@ def _assert(condition, message):
...
-import torch.testing
-import torch.backends.cuda
-import torch.backends.mkl
-import torch.backends.mkldnn
-import torch.backends.openmp
-import torch.backends.quantized
-import torch.quantization
-import torch.utils.data
 import torch.__config__
 import torch.__future__
```
Comment (jakebailey) [inline torch/__init__.py:606]: Do these need to be exported? These appear to be unused in in this file.

File: torch/__init__.py [hunk 7]
```diff
@@ -616,7 +624,7 @@ def compiled_with_cxx11_abi():
 from torch._classes import classes
```
Comment (jakebailey) [inline torch/__init__.py:624]: This is also unused; maybe worthwhile checking each of these to see if there are any unused.

File: torch/nn/__init__.py [hunk 8]
```diff
@@ -3,3 +3,9 @@
 from .parallel import DataParallel
 from . import init
 from . import utils
+
+# Re-export submodules
+
+from torch.nn import intrinsic as intrinsic
+from torch.nn import quantizable as quantizable
+from torch.nn import quantized as quantized
```
Comment (jakebailey) [inline torch/nn/__init__.py:11]: Should these be `from .nn` to match the above?

File: torch/nn/__init__.py [hunk 9]
```diff
@@ -3,3 +3,9 @@
 from .parallel import DataParallel
 from . import init
 from . import utils
```
Comment (jakebailey) [inline torch/nn/__init__.py:5]: These lines above (github won't let me select them) appear to also only exist to reexport things, but I'm not entirely certain as an outsider... 🙂

File: torch/nn/__init__.py (hunk 8 above)
Comment (jakebailey) [reply torch/nn/__init__.py:11]: (I don't know if there's a semantic difference off the top of my head.)

**Comment by jakebailey** [comment]
I'm wondering if you could speed this up by opening the repo in something like Pylance (or the pyright CLI) to see which imports are dead code. IIRC it should also inform about import cycles.

**Comment by jakebailey** [comment]
@MHDante Are you still planning on working on this change? I'd be happy to look at some of the failures here to try and help get this unblocked (or take it over if you're no longer interested). I'm certainly interested in getting this fixed as it'll really help editor experiences.

**Comment by MHDante** [comment]
Oof, yeah, I've been a bit slammed at work. I'll give it a shot this weekend. Apologies for the delay. 

@jakebailey  If you want, I've added you to [this PR's fork ](https://github.com/MHDante/pytorch), so feel free to push changes there.

File: torch/nn/__init__.py (hunk 8 above)
Comment (gramster) [reply torch/nn/__init__.py:11]: From `.nn` seems safer to me but IDK either; @brettcannon, is there a semantic difference?

Comment (brettcannon) [reply torch/nn/__init__.py:11]: It's essentially stylistic/preference; I personally prefer relative imports as it makes it clearer that the module is from the same package and renames won't affect anything. But it does make relocating the file in the future slightly harder.

**Comment by jakebailey** [comment]
Thanks for that access (meant to reply sooner). Note that you are committing directly to your `master` branch which may make it a little difficult to fix up the merge conflicts and keep the fork "clean". Perhaps it'd be worth moving the changes to a branch and resubmitting (I don't know if GH lets you change the PR's source).

**Comment by jakebailey** [comment]
FWIW I've opened #52339 which does a small subset of these changes; I was having quite a bit of trouble trying to apply the rules to every single module and the bulk of the problems I've personally seen are just those in `torch/__init__.py`.

**Comment by MHDante** [comment]
Yeah. I think I will close this PR. It seems a bit involved, and it might be a better approach to do it a bit at a time.

Apologies for the holdup.

**Comment by jakebailey** [comment]
No worries, I really appreciate the effort!

**Pull request closed without merging**